### PR TITLE
raft: add match index safety check on followers

### PR DESCRIPTION
**Pull request by pav-kv**
This PR adds a safety check which ensures that a follower's log state does not contradict to the leader's `Match` index.

If the leader has a non-zero `Match` for the follower, it believes that the follower's log a) is consistent with this leader, and b) has all entries up to the `Match` index persisted. If the follower's `Term <= leader.Term` and (b) turns out to not be the case, this means the follower has lost part of the durable state (for example, fsync didn't work properly in the system, and the node got restarted). This is a safety violation, so this follower panics to avoid spreading the harm.

In the future, it would be better to "quarantine" this follower, and surface this information to the operator without crashing the node.

Touches #122690
Epic: none
Release note: none

### CODE REVIEW

File: pkg/raft/raft.go [hunk 1]
```diff
@@ -1702,7 +1706,22 @@ func (r *raft) handleAppendEntries(m pb.Message) {
...
+// leader's idea where it matches.
+func (r *raft) checkMatch(match uint64) {
+	// TODO(pav-kv): lastIndex() might be not yet durable. Make this check
+	// stronger by comparing `match` with the last durable index.
+	//
+	// TODO(pav-kv): make this check stronger when the raftLog stores the last
+	// accepted term. If `match` is non-zero, this follower's log last accepted
+	// term must equal the leader term, and have entries up to `match` durable.
+	if last := r.raftLog.lastIndex(); last < match {
+		r.logger.Panicf("match(%d) is out of range [lastIndex(%d)]. Was the raft log corrupted, truncated, or lost?", match, last)
```
Comment (nvanbenschoten) [inline pkg/raft/raft.go:1719]: Should we add a test for the MsgApp and MsgHeartbeat cases?

**Review by nvb — APPROVED** [review]
I think this makes sense. It's a more direct version of the commit index regression check, so it unblocks https://github.com/cockroachdb/cockroach/pull/122690. It also sets us up well to do more in this area in the future while removing our dependency on `MsgHeartbeat`.

File: pkg/raft/raft.go (hunk 1 above)
Comment (pav-kv) [reply pkg/raft/raft.go:1719]: Done.

**Comment by pav-kv** [comment]
bors r=nvanbenschoten

**Comment by pav-kv** [comment]
bors cancel

**Comment by pav-kv** [comment]
bors retry

**Comment by pav-kv** [comment]
bors cancel

**Pull request merged**